pm2 can't startup in osx

I want to run pm2 when osx startup, so I follow this [doc](http://pm2.keymetrics.io/docs/usage/startup/).

#### Steps to reproduce
```
pm2 startup
PM2] Init System found: launchd
[PM2] You have to run this command as root. Execute the following command:
sudo env PATH=$PATH:/Users/google/.tnvm/versions/alinode/v2.0.5/bin /Users/google/.tnvm/versions/alinode/v2.0.5/lib/node_modules/pm2/bin/pm2 startup launchd -u google --hp /Users/google
```
then I run the command, but error occurs, here is log:
```
PM2] Init System found: launchd
Platform launchd
Template
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
  <dict>
	  <key>Label</key>
	  <string>com.PM2</string>
	  <key>UserName</key>
	  <string>google</string>
    <key>KeepAlive</key>
    <true/>
	  <key>ProgramArguments</key>
	  <array>
		  <string>/bin/sh</string>
		  <string>-c</string>
		  <string>/Users/google/.tnvm/versions/alinode/v2.0.5/lib/node_modules/pm2/bin/pm2 resurrect</string>
	  </array>
	  <key>RunAtLoad</key>
	  <true/>
	  <key>OnDemand</key>
	  <false/>
	  <key>LaunchOnlyOnce</key>
	  <true/>
	  <key>EnvironmentVariables</key>
    <dict>
      <key>PATH</key>
      <string>/Users/google/.tnvm/versions/alinode/v2.0.5/bin</string>
      <key>PM2_HOME</key>
      <string>/Users/google/.pm2</string>
    </dict>
	  <key>StandardErrorPath</key>
	  <string>/tmp/com.PM2.err</string>
	  <key>StandardOutPath</key>
	  <string>/tmp/com.PM2.out</string>
  </dict>
</plist>

Target path
/Users/google/Library/LaunchAgents/PM2.plist
Command list
[ 'chmod +x /Users/google/Library/LaunchAgents/PM2.plist',
  'launchctl load /Users/google/Library/LaunchAgents/PM2.plist',
  'launchctl start /Users/google/Library/LaunchAgents/PM2.plist',
  'launchctl list' ]
[PM2] Writing init configuration in /Users/google/Library/LaunchAgents/PM2.plist
[PM2] Making script booting at startup...
>>> Executing chmod +x /Users/google/Library/LaunchAgents/PM2.plist
[DONE] 
>>> Executing launchctl load /Users/google/Library/LaunchAgents/PM2.plist
[DONE] 
>>> Executing launchctl start /Users/google/Library/LaunchAgents/PM2.plist
[ERROR] Exit code : 3
[PM2][ERROR] launchctl start /Users/google/Library/LaunchAgents/PM2.plist failed, see error above.
```

#### Software versions used

```
OS         :10.11.6
node.js    :v6.9.2
PM2        :2.2.3
```

I installed node with brew & pm2 with npm.

The following file exists:

 /usr/local/lib/node_modules/pm2/bin/pm2

Contents /tmp/com.PM2.err:

[PM2][ERROR] No processes saved; DUMP file doesn't exist

Contents /tmp/com.PM2.err (portals is my app):

[PM2] Resurrecting
[PM2] Restoring processes located in /Users/channel1/.pm2/dump.pm2
┌──────────┬────┬──────┬──────┬────────┬─────────┬────────┬─────┬───────────┬──────────┐
│ App name │ id │ mode │ pid  │ status │ restart │ uptime │ cpu │ mem       │ watching │
├──────────┼────┼──────┼──────┼────────┼─────────┼────────┼─────┼───────────┼──────────┤
│ portals  │ 0  │ fork │ 4647 │ online │ 0       │ 25h    │ 0%  │ 47.7 MB   │ enabled  │
└──────────┴────┴──────┴──────┴────────┴─────────┴────────┴─────┴───────────┴──────────┘
 Use `pm2 show <id|name>` to get more details about an app






On Jan 27, 2017, at 12:42 PM, Antoine Bluchet <notifications@github.com<mailto:notifications@github.com>> wrote:


Hmm, this seems wrong, but may still be correct:

          <key>ProgramArguments</key>
          <array>
                  <string>/bin/sh</string>
                  <string>-c</string>
                  <string>/usr/local/lib/node_modules/pm2/bin/pm2 resurrect</string>
          </array>


Did you install nodejs with brew? Does the binary /usr/local/lib/node_modules/pm2/bin/pm2 exists?

After running the launchctl start ~/Library/LaunchAgents/PM2.plist command, is there anything in /tmp/com.PM2.err or /tmp/com.PM2.out?

Thanks!

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub<https://github.com/Unitech/pm2/issues/2617#issuecomment-275726182>, or mute the thread<https://github.com/notifications/unsubscribe-auth/ACWWc67fngFsOMhR6kpw_1L-npGJSOmFks5rWix_gaJpZM4Ld8fe>.

--- a/lib/API/Startup.js
+++ b/lib/API/Startup.js
@@ -174,7 +174,7 @@ module.exports = function(CLI) {
     }
     if (platform == null)
       throw new Error('Init system not found');
-    
+
     if (!cb) {
       cb = function(err, data) {
         if (err)
@@ -240,14 +240,10 @@ module.exports = function(CLI) {
     case 'macos':
     case 'darwin':
     case 'launchd':
-      // https://nathangrigg.com/2012/07/schedule-jobs-using-launchd#launchctl
       template = getTemplate('launchd');
       destination = path.join(process.env.HOME, 'Library/LaunchAgents/PM2.plist');
       commands = [
-        'chmod +x ' + destination,
-        'launchctl load ' + destination,
-        'launchctl start ' + destination,
-        'launchctl list'
+        'launchctl load -w ' + destination
       ]
       break;
     case 'freebsd':

